### PR TITLE
feat: grouped closed-loop eval as ClosedLoopEvaluation subclass (PR 5/6)

### PR DESCRIPTION
**Pull request by Owen-Liuyuxuan**
## Summary

Split stack **5/6** for [tier4/Diffusion-Planner#205](https://github.com/tier4/Diffusion-Planner/pull/205). **Base:** PR-4.

Grouped-by-area closed-loop validation as **`GroupedByAreaClosedLoopEvaluation`** — subclasses `ClosedLoopEvaluation` and overrides only discovery, job execution, summary, artifacts, and DDP shard extras.

### CLI (unified)

```python
evaluator = create_evaluator_from_args(args)  # FullRoute or GroupedClassificationEvalSuite
summary = evaluator.run()
```

No `run_grouped_eval` / `run_full_route_eval` workflow split. Multi-date jobs use **`GroupedClassificationEvalSuite`**.

### Rollout stack

```text
run_bag_rollout → render_segment(instrument=True)
  └─ metrics/grouped_by_area.py (post-process only)
```

## Key files

- `grouped_closed_loop_eval.py` — `GroupedByAreaClosedLoopEvaluation`, `GroupedClassificationEvalSuite`
- `scenario_classification.py` — JSON discovery + merge
- `metrics/*` — per-area metric plugins
- `closed_loop_rollout.py` — shared bag rollout
- `wandb_closed_loop.py` — training log builders

## Test plan

- [ ] `valid_predictor_closed_loop.py --mode grouped ...`
- [ ] `tools/eval_scenario_groups_closed_loop.py`
- [ ] DDP bag sharding via `GroupedByAreaClosedLoopEvaluation.run_distributed()`
- [ ] Output: `results_table.csv`, `by_area/`, `videos/`, `summary.json`

## Notes for reviewers

- [ ] No backward-compat shims (`run_grouped_closed_loop_eval`, re-export modules)
- [ ] No `closed_loop_runner.py` — workflow stays in evaluator classes
- [ ] Grouped path does not duplicate segment loop outside `render_segment`

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by Owen-Liuyuxuan** [comment]
**Rebased/refactored** after review feedback:

- Removed `closed_loop_rollout.py` — instrumented bag rollout is inlined in `GroupedByAreaClosedLoopEvaluation._rollout_bag()` / `run_job()`.
- Removed `GroupedClassificationEvalSuite` — multi-date discovery is in `GroupedMultiDateClosedLoopEvaluation.discover_jobs()`.
- **No changes** to `closed_loop_evaluation.py` / DDP (those stay in #233).
- Added `ResolvedClosedLoopEvaluation` for training (#235): grouped vs full-route resolution inside the class, not in `train.py`.

File: scenario_generation/closed_loop_types.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    k: int
+    rec_idx: int
+    area: str | None
+    centerline_m: float | None = None
+    turn_match: bool | None = None
+    clearance_m: float = float("inf")
+    collision: bool = False
+    neighbor_violation: bool = False
+    rb_violation: bool = False
```
Comment (kosuke55) [inline scenario_generation/closed_loop_types.py:29]:
```suggestion
    rb_violation: bool = False
    rb_dist_m: float = float("inf")
```

File: scenario_generation/reproducer_rollout.py [hunk 2]
```diff
@@ -1541,27 +1560,66 @@ def render_segment(
...
+                StepRecord(
+                    k=int(k),
+                    rec_idx=int(idx),
+                    area=area,
+                    centerline_m=centerline_m,
+                    turn_match=turn_match,
+                    clearance_m=float(safety["clearance_m"]),
+                    collision=bool(safety["collision"]),
+                    neighbor_violation=bool(safety["neighbor_violation"]),
+                    rb_violation=bool(safety["rb_violation"]),
```
Comment (kosuke55) [inline scenario_generation/reproducer_rollout.py:1608]:
```suggestion
                    rb_violation=bool(safety["rb_violation"]),
                    rb_dist_m=float(safety["rb_dist_m"]),
```

File: scenario_generation/metrics/safety_clearance.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        ego_shape_t = torch.tensor(
+            np.asarray(np_dict["ego_shape"]).reshape(-1)[:3],
+            dtype=torch.float32,
+            device=device,
+        )
+        traj = torch.zeros(1, 1, 4, dtype=torch.float32, device=device)
+        traj[0, 0, 2] = 1.0  # heading +x in ego frame
+        data = {
+            "line_strings": torch.tensor(
+                np_dict["line_strings"][None], dtype=torch.float32, device=device
```
Comment (kosuke55) [inline scenario_generation/metrics/safety_clearance.py]:
```suggestion
                np_dict["line_strings"], dtype=torch.float32, device=device
```

File: scenario_generation/metrics/grouped_by_area.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        "centerline_p95_m": _percentile(cl, 95) if cl.size else None,
+        "centerline_max_m": float(cl[np.isfinite(cl)].max()) if cl.size else None,
+        "turn_match_rate": float(tm.mean()) if tm.size else None,
+        "neighbor_violation_steps": int(sum(st.neighbor_violation for st in area_steps)),
+        "rb_violation_steps": int(sum(st.rb_violation for st in area_steps)),
+        "collision_steps": int(sum(collisions)),
+        "min_clearance_m": float(clearances[np.isfinite(clearances)].min())
+        if clearances.size
+        else float("inf"),
+        "min_rb_dist_m": None,
```
Comment (kosuke55) [inline scenario_generation/metrics/grouped_by_area.py]:
```suggestion
        rb_dists = np.array([st.rb_dist_m for st in area_steps], dtype=np.float32)
        "min_rb_dist_m": float(rb_dists[np.isfinite(rb_dists)].min()) if rb_dists.size and np.isfinite(rb_dists).any() else None,
```

**Review by kosuke55 — COMMENTED** [review]
`min_rb_dist_m` was always None.
Comfirmed that the valued can be calculed with these fixes.

image

File: scenario_generation/metrics/grouped_by_area.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        "sequence": route_key,
+        "bag": bag_name,
+        "span_index": span_index,
+        "list_start_idx": video_start_idx,
+        "list_end_idx": video_end_idx,
+        "labeled_ranges": labeled_ranges,
+        "segment": f"[{f_start},{f_end}]",
+        "n_steps_run": len(area_steps),
+        "terminated": "area_span",
+        "min_clearance": float(clearances[np.isfinite(clearances)].min())
```
Comment (TatsukiNishimura) [inline scenario_generation/metrics/grouped_by_area.py]: `clearances` can be `[inf, inf, ...]` When `area_steps` is non-empty and no valid neighbor exists at any step. In that case, `clearances.size` is non-zero, but
  `clearances[np.isfinite(clearances)]` is empty, so `.min()` / `.mean()` raises a `ValueError`.

  Could we check the size of the finite subset instead and return `inf` when it is empty?

File: scenario_generation/metrics/centerline_deviation.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+def _valid_centerline_points(route_lanes: np.ndarray) -> np.ndarray:
+    valid = np.abs(route_lanes[..., :2]).sum(axis=-1) > 1e-6
+    pts = route_lanes[..., :2][valid]
+    if pts.size == 0:
+        return np.zeros((0, 2), dtype=np.float32)
+    return pts.reshape(-1, 2).astype(np.float32)
+
+
+def lateral_offset_from_route_lanes(route_lanes: np.ndarray) -> float:
```
Comment (TatsukiNishimura) [inline scenario_generation/metrics/centerline_deviation.py]: This computes the minimum Euclidean distance from the ego origin to sampled
centerline points, rather than the lateral offset from the centerline.
For example, a centerline point at (10, 0) gives 10 m even when the ego is
perfectly centered. Should this project the ego position onto the nearest
centerline segment and measure the perpendicular distance instead?

Comment (Owen-Liuyuxuan) [reply scenario_generation/metrics/centerline_deviation.py]: @TatsukiNishimura I decided to rename the code into computing centerline point distance. I try to avoid making too huge refactoring on the current PR. For now, the centerline distance is on the point distance -> which will be the most simple one https://github.com/tier4/Diffusion-Planner/pull/234/commits/187eeac04d37dd160e0312cb2b4cf4963fd9c39a

File: scenario_generation/metrics/grouped_by_area.py (hunk 5 above)
Comment (Owen-Liuyuxuan) [reply scenario_generation/metrics/grouped_by_area.py]: @TatsukiNishimura  Thanks for finding the bug.

https://github.com/tier4/Diffusion-Planner/pull/234/commits/187eeac04d37dd160e0312cb2b4cf4963fd9c39a

File: scenario_generation/metrics/grouped_by_area.py (hunk 4 above)
Comment (Owen-Liuyuxuan) [reply scenario_generation/metrics/grouped_by_area.py]: @kosuke55  these comments are applied.
